animation_nodes/nodes/falloff/mesh_falloff_utils.pyx

```diff
@@ -0,0 +1,130 @@
+        self.useVolume = useVolumeInput
+        self.invert = invertInput
+        if falloffWidth < 0:
+            size += falloffWidth
+            falloffWidth = -falloffWidth
+        self.minDistance = size
+        self.maxDistance = size + falloffWidth
+
+        if self.minDistance == self.maxDistance:
+            self.minDistance -= 0.00001
+        self.factor = 1 / (self.maxDistance - self.minDistance)
+
+        self.dataType = "LOCATION"
+        self.clamped = True
+
+    cdef float evaluate(self, void *value, Py_ssize_t index):
+        cdef float distance
+        cdef float strength
+        cdef Vector3* v
+        if self.useVolume:
+            v = <Vector3*>value
+            strength = hitsPerLocation(self.bvhTree, Vector((v.x, v.y, v.z)))
+            distance = calculateDistance(self, v)
+            strength = max(distance, strength)
+            if self.invert: strength = 1 - strength
+            return strength
+        else:
+            strength = calculateDistance(self, <Vector3*>value)
+            if self.invert: strength = 1 - strength
+            return strength
+
+    cdef void evaluateList(self, void *values, Py_ssize_t startIndex, Py_ssize_t amount, float *target):
+        cdef float distance
+        cdef float strength
+        cdef Vector3* v
+        cdef Py_ssize_t i
+
+        if self.useVolume:
+            for i in range(amount):
+                v = <Vector3*>values + i
+                strength = hitsPerLocation(self.bvhTree, Vector((v.x, v.y, v.z)))
+                distance = calculateDistance(self, v)
+                strength = max(distance, strength)
+                if self.invert: strength = 1 - strength
+                target[i] = strength
+        else:
+            for i in range(amount):
+                strength = calculateDistance(self, <Vector3*>values + i)
+                if self.invert: strength = 1 - strength
+                target[i] = strength
+
+cdef inline float calculateDistance(calculateMeshSurfaceFalloff self, Vector3 *v):
+    cdef float distance = self.bvhTree.find_nearest(Vector((v.x, v.y, v.z)), self.bvhMaxDistance)[3]
+    if distance <= self.minDistance: return 1
+    if distance <= self.maxDistance: return 1 - (distance - self.minDistance) * self.factor
+    return 0
+
+
+cdef class calculateMeshVolumeFalloff(BaseFalloff):
+    cdef:
+        bvhTree
+        bint invert
+
+    def __cinit__(self, bvhTreeInput, bint invertInput):
+        self.bvhTree = bvhTreeInput
+        self.invert = invertInput
+
+        self.dataType = "LOCATION"
+        self.clamped = True
+
+    cdef float evaluate(self, void *value, Py_ssize_t index):
+        return calculateStrength(self, <Vector3*>value)
+
+    cdef void evaluateList(self, void *values, Py_ssize_t startIndex, Py_ssize_t amount, float *target):
+        cdef Py_ssize_t i
+        for i in range(amount):
+            target[i] = calculateStrength(self, <Vector3*>values + i)
```

Comment:
The evaluateList method doesn't seem to be needed in this case.

Reply:
Omar, Can you please elaborate why we do not need the `evaluateList` method here?

Reply:
The base falloff implement the `evaluateList` method:

```
    cdef void evaluateList(self, void *objects, Py_ssize_t startIndex,
                           Py_ssize_t amount, float *target):
        cdef Py_ssize_t i
        cdef Py_ssize_t elementSize = getSizeOfFalloffDataType(self.dataType)
        for i in range(amount):
            target[i] = self.evaluate(<char*>objects + i * elementSize, i + startIndex)
```

In this case, your overriding method doesn't seem to provide any gain, so we can just leave it as is.

Reply:
@3DSinghVFX It should be noted that your new version of the code could use an `evaluateList`  method to avoid all of the unnecessary branching and checks.